pkg/network/usm/sharedlibraries/ebpf.go

```diff
@@ -0,0 +1,207 @@
+
+//go:build linux_bpf
+
+package sharedlibraries
+
+import (
+	"fmt"
+	"math"
+	"os"
+
+	manager "github.com/DataDog/ebpf-manager"
+	"golang.org/x/sys/unix"
+
+	ddebpf "github.com/DataDog/datadog-agent/pkg/ebpf"
+	"github.com/DataDog/datadog-agent/pkg/ebpf/bytecode"
+	"github.com/DataDog/datadog-agent/pkg/network/config"
+	netebpf "github.com/DataDog/datadog-agent/pkg/network/ebpf"
+	errtelemetry "github.com/DataDog/datadog-agent/pkg/network/telemetry"
+	"github.com/DataDog/datadog-agent/pkg/util/kernel"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+const (
+	maxActive              = 1024
+	sharedLibrariesPerfMap = "shared_libraries"
+	probeUID               = "so"
+
+	// probe used for streaming shared library events
+	openatSysCall  = "openat"
+	openat2SysCall = "openat2"
+)
+
+var traceTypes = []string{"enter", "exit"}
+
+type ebpfProgram struct {
+	cfg         *config.Config
+	probeIDs    []manager.ProbeIdentificationPair
+	perfHandler *ddebpf.PerfHandler
+	*errtelemetry.Manager
+}
+
+func newEBPFProgram(c *config.Config, bpfTelemetry *errtelemetry.EBPFTelemetry) *ebpfProgram {
+	perfHandler := ddebpf.NewPerfHandler(100)
+	mgr := &manager.Manager{
+		PerfMaps: []*manager.PerfMap{
+			{
+				Map: manager.Map{
+					Name: sharedLibrariesPerfMap,
+				},
+				PerfMapOptions: manager.PerfMapOptions{
+					PerfRingBufferSize: 8 * os.Getpagesize(),
+					Watermark:          1,
+					RecordHandler:      perfHandler.RecordHandler,
+					LostHandler:        perfHandler.LostHandler,
+					RecordGetter:       perfHandler.RecordGetter,
+				},
+			},
+		},
+	}
+
+	probeIDs := getSysOpenHooksIdentifiers()
+	for _, identifier := range probeIDs {
+		mgr.Probes = append(mgr.Probes,
+			&manager.Probe{
+				ProbeIdentificationPair: identifier,
+				KProbeMaxActive:         maxActive,
+			},
+		)
+	}
+
+	return &ebpfProgram{
+		cfg:         c,
+		Manager:     errtelemetry.NewManager(mgr, bpfTelemetry),
+		probeIDs:    probeIDs,
+		perfHandler: perfHandler,
+	}
+}
+
+func (e *ebpfProgram) Init() error {
+	e.InstructionPatcher = func(m *manager.Manager) error {
+		return errtelemetry.PatchEBPFTelemetry(m, true, nil)
+	}
+
+	var err error
+	if e.cfg.EnableCORE {
+		err = e.initCORE()
+		if err == nil {
+			return nil
+		}
+
+		if !e.cfg.AllowRuntimeCompiledFallback && !e.cfg.AllowPrecompiledFallback {
+			return fmt.Errorf("co-re load failed: %w", err)
+		}
+		log.Warnf("co-re load failed. attempting fallback: %s", err)
+	}
+
+	if e.cfg.EnableRuntimeCompiler || (err != nil && e.cfg.AllowRuntimeCompiledFallback) {
+		err = e.initRuntimeCompiler()
+		if err == nil {
+			return nil
+		}
+
+		if !e.cfg.AllowPrecompiledFallback {
+			return fmt.Errorf("runtime compilation failed: %w", err)
+		}
+		log.Warnf("runtime compilation failed: attempting fallback: %s", err)
+	}
+
+	return e.initPrebuilt()
+}
+
+func (e *ebpfProgram) GetPerfHandler() *ddebpf.PerfHandler {
+	return e.perfHandler
+}
+
+func (e *ebpfProgram) Stop() {
```

Comment:
don't we need to iterate over the probes and stop them?
(like the code that you removed from `ebpf_ssl.Stop()`)

Reply:
Good question – I will have to take a look at the `Manager` code. I don't think we do because differently from the old code where we were only stopping a subset of probes, here we're stopping the whole eBPF "program" by calling `Manager.Stop`

Reply:
IIRC, the manager Stop All stopped everything for us
The reason I made the old version was due to the fact we had a  single manager, and we needed to terminate and stop probes and the perf before we have the chance to call Stop(ALL)

Reply:
I just confirmed that the Manager indeed stops all probes, so nothing to do here